Core/include/Acts/SpacePointFormation/SingleHitSpacePointBuilder.hpp

```diff
@@ -0,0 +1,69 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2018-2021 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+#include "Acts/Definitions/TrackParametrization.hpp"
+#include "Acts/Geometry/GeometryContext.hpp"
+#include "Acts/Geometry/TrackingGeometry.hpp"
+#include "Acts/SpacePointFormation/SpacePointBuilderConfig.h"
+#include "Acts/Surfaces/Surface.hpp"
+
+namespace Acts {
+
+/// @class SingleHitSpacePointBuilder
+///
+/// After the particle interaction with surfaces are recorded and digitized
+/// the clusters pixel detectors need further treatment. This class takes
+/// the digitized clusters on a pixel detector element and provides the
+/// corresponding space point.
+///
+template <typename spacepoint_t, typename cluster_t>
+class SingleHitSpacePointBuilder {
+ public:
+  struct Config {
+    // Tracking geometry for transformation lookup.
+    std::shared_ptr<const Acts::TrackingGeometry> trackingGeometry;
+  };
+
+  // Constructor
+  SingleHitSpacePointBuilder(SingleHitSpacePointBuilderConfig cfg);
+  ///// Default constructor
+  SingleHitSpacePointBuilder() = default;
+
+  /// @brief Calculates the space points out of a given collection of clusters
+  /// and stores the results
+  ///
+  /// @param gctx The current geometry context object, e.g. alignment
+  /// @param clusters vector of clusters
+  /// @param spacePointStorage storage of the results
+  void calculateSpacePoints(const GeometryContext& gctx,
+                            const std::vector<cluster_t>& clusters,
+                            std::vector<spacepoint_t>& spacePointStorage) const;
```

Comment:
What about using iterators for the input collection and std::back_insert_iterator for the output collection? Would this make sense?

Something such as 
```
template<template <typename...> typename container_t, typename iterator_t>
void calculateSpacePoints(const GeometryContext& gctx,
                            const iterator_t& clusters_begin,
                            const iterator_t& clusters_end,
                            std::back_insert_iterator<container_t<spacepoint_t>>& spacePointStorage) const;
```